File: Módulo5 - WebServices/WS05 - Valida CPF (Varios Provedores)/ws provider/provider.cpp
//curl -L -o httplib.h https://raw.githubusercontent.com/yhirose/cpp-httplib/master/httplib.h
//g++ -std=c++11 -o provider provider.cpp -pthread
//./provider
#include <iostream>
#include <string>
#include <algorithm>
#include <sstream>
#include <map>
#include "httplib.h"

// Serviço fornecido pelo Web Service Provider
class CPFService {
public:
    bool validarCpf(const std::string& cpf) {
        // Remove todos os caracteres não numéricos do CPF, mantendo apenas dígitos
        std::string cpf_limpo = cpf;
        cpf_limpo.erase(std::remove(cpf_limpo.begin(), cpf_limpo.end(), '.'), cpf_limpo.end());
        cpf_limpo.erase(std::remove(cpf_limpo.begin(), cpf_limpo.end(), '-'), cpf_limpo.end());
        
        // Verifica se o CPF tem exatamente 11 dígitos após a limpeza
        if (cpf_limpo.length() != 11) {
            return false;
        }
        
        // Verifica se todos os dígitos são iguais (ex: 11111111111, 22222222222)
        bool todos_iguais = true;
        for (size_t i = 1; i < cpf_limpo.length(); i++) {
            if (cpf_limpo[i] != cpf_limpo[0]) {
                todos_iguais = false;
                break;
            }
        }
        if (todos_iguais) {
            return false;
        }

        // Calcula os dígitos verificadores
        int dig1 = calcDigito(cpf_limpo, 10);
        int dig2 = calcDigito(cpf_limpo, 11);

        // Compara os dois últimos dígitos do CPF com os dígitos calculados
        std::string digitos_calculados = std::to_string(dig1) + std::to_string(dig2);
        std::string digitos_cpf = cpf_limpo.substr(9, 2);
        
        return digitos_cpf == digitos_calculados;
    }

private:
    // Define função interna para calcular um dígito verificador do CPF
    int calcDigito(const std::string& cpf, int peso) {
        int s = 0;
        for (int i = 0; i < peso - 1; i++) {
            s += (cpf[i] - '0') * (peso - i);
        }
        int resto = (s * 10) % 11;
        if (resto == 10) {
            return 0;
        } else {
            return resto;
        }
    }
};

// Função para criar resposta JSON
std::string criarJsonResposta(const std::string& cpf, bool valido) {
    std::ostringstream json;
    json << "{\"cpf\":\"" << cpf << "\",\"valido\":" << (valido ? "true" : "false") << "}";
    return json.str();
}

std::string criarJsonErro(const std::string& mensagem) {
    std::ostringstream json;
    json << "{\"error\":\"" << mensagem << "\"}";
    return json.str();
}

int main() {
    httplib::Server svr;
    CPFService cpfService;

    // Handler para requisições GET /validar
    svr.Get("/validar", [&cpfService](const httplib::Request& req, httplib::Response& res) {
        std::cout << "Requisição GET recebida - Validar CPF" << std::endl;
        
        // Obtém o parâmetro cpf da query string
        auto it = req.params.find("cpf");
        if (it == req.params.end()) {
            res.status = 400;
            res.set_content(criarJsonErro("Parâmetro 'cpf' não fornecido"), "application/json");
            return;
        }
        
        std::string cpf = it->second;
        
        // Realiza a operação de validação
        bool resultado = cpfService.validarCpf(cpf);
        
        // Envia a resposta para o cliente
        res.status = 200;
        res.set_content(criarJsonResposta(cpf, resultado), "application/json");
    });

    // Handler para requisições POST
    svr.Post("/.*", [](const httplib::Request& req, httplib::Response& res) {
        // Este provider só aceita requisições GET
        res.status = 405;  // 405 Method Not Allowed
        res.set_header("Allow", "GET");  // Informa que apenas GET é permitido
        res.set_content(criarJsonErro("Método HTTP 'POST' não é permitido. Apenas o método GET é suportado para este endpoint."), "application/json");
    });

    // Handler catch-all para retornar 404 quando o endpoint não é /validar
    // Este handler só será chamado se nenhum handler anterior corresponder
    svr.Get(".*", [](const httplib::Request& req, httplib::Response& res) {
        // Verifica se o path não começa com /validar (para evitar conflito com o handler específico)
        if (req.path.find("/validar") != 0) {
            res.status = 404;
            res.set_content(criarJsonErro("Endpoint não encontrado"), "application/json");
        }
    });

    // Handler genérico para retornar 404 em caso de erro
    svr.set_error_handler([](const httplib::Request& req, httplib::Response& res) {
        if (res.status == 404) {
            res.set_content(criarJsonErro("Endpoint não encontrado"), "application/json");
        }
    });

    // Inicia o servidor HTTP na porta 8084
    std::cout << "Servidor iniciado em http://127.0.0.1:8084" << std::endl;
    svr.listen("127.0.0.1", 8084);

    return 0;
}

File: Módulo5 - WebServices/WS05 - Valida CPF (Varios Provedores)/ws provider/provider.py
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

# Classe Provider para lidar com as requisições HTTP
class WSProvider(BaseHTTPRequestHandler):

    def do_GET(self):
        parsed_path = urlparse(self.path)    
        # Obtém os parâmetros da requisição GET
        query_params = parse_qs(parsed_path.query)
        
        # Verifica se o parâmetro cpf está presente
        if query_params["cpf"]:
            cpf = query_params["cpf"][0]
        else:
            self.send_response(400)  # Bad Request
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            resposta = json.dumps({"error": "Parâmetro 'cpf' não fornecido"}).encode('utf-8')
            self.wfile.write(resposta)
            return
        

        if self.path.startswith("/validar"):           
            print("Requisição GET recebida - Validar CPF")
            # Realiza a operação de validação
            valida = CPFService()
            resultado = valida.validarCpf(cpf)
        else:
            # Retorna erro 404 quando o endpoint não é /validar
            self.send_response(404)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            resposta = json.dumps({"error": "Endpoint não encontrado"}).encode('utf-8')
            self.wfile.write(resposta)
            return

        # Envia a resposta para o cliente
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()
        msg_retorno = {"cpf": cpf, "valido": resultado}
        resposta = json.dumps(msg_retorno).encode('utf-8')
        self.wfile.write(resposta)

    def do_POST(self):
        # Este provider só aceita requisições GET
        self.send_response(405)  # 405 Method Not Allowed
        self.send_header('Content-type', 'application/json')
        self.send_header('Allow', 'GET')  # Informa que apenas GET é permitido
        self.end_headers()
        resposta = json.dumps({"error": "Método HTTP 'POST' não é permitido. Apenas o método GET é suportado para este endpoint."}).encode('utf-8')
        self.wfile.write(resposta)
    
# Serviço fornecido pelo Web Service Provider
class CPFService:
    def validarCpf(self, cpf: str) -> bool:
        # Remove todos os caracteres não numéricos do CPF, mantendo apenas dígitos
        cpf_limpo = cpf.replace(".", "").replace("-", "") 
        # Verifica se o CPF tem exatamente 11 dígitos após a limpeza
        if len(cpf_limpo) != 11:
            return False
        # Verifica se todos os dígitos são iguais (ex: 11111111111, 22222222222)
        if cpf_limpo == cpf_limpo[0] * 11:
            return False

        # Calcula os dígitos verificadores
        dig1 = self.__calcDigito(cpf_limpo, 10)
        dig2 = self.__calcDigito(cpf_limpo, 11)

        # Compara os dois últimos dígitos do CPF com os dígitos calculados
        return cpf_limpo[-2:] == f"{dig1}{dig2}"

    # Define função interna para calcular um dígito verificador do CPF
    def __calcDigito(self, cpf, peso):
        s = 0
        for i in range(peso - 1):
            s += int(cpf[i]) * (peso - i)
        resto = (s * 10) % 11
        if resto == 10:
            return 0
        else:
            return resto

def main():
    # Inicia o servidor HTTP na porta 8080
    servidor = HTTPServer(('127.0.0.1', 8080), WSProvider)
    print("Servidor iniciado em http://127.0.0.1:8080")
    servidor.serve_forever()

if __name__ == "__main__":
    main()
